experimental/codepointtrie/src/codepointtrie.rs

```diff
@@ -0,0 +1,311 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+// TODO: add module-level Rust-doc with examples
+
+use crate::error::Error;
+use crate::impl_const::*;
+use std::marker::PhantomData;
+use zerovec::ZeroVec;
+
+// Enums
+
+/// The width of the elements in the data array of a [`CodePointTrie`].
+/// See UCPTrieValueWidth in ICU4C.
+#[derive(Clone, Copy, PartialEq)]
+pub enum ValueWidthEnum {
+    Bits16 = 0,
+    Bits32 = 1,
+    Bits8 = 2,
+}
+
+/// The type of trie represents whether the trie has an optimization that
+/// would make it small or fast.
+/// See UCPTrieType in ICU4C.
+#[derive(Clone, Copy, PartialEq)]
+pub enum TrieTypeEnum {
+    Fast = 0,
+    Small = 1,
+}
+
+// ValueWidth trait
+
+// AsULE is AsUnalignedLittleEndian, i.e. "allowed in a zerovec"
+
+/// A trait representing the width of the values stored in the data array of a
+/// [`CodePointTrie`]. This trait is used as a type parameter in constructing
+/// a `CodePointTrie`.
+pub trait ValueWidth: Copy + zerovec::ule::AsULE {
+    /// This enum variant represents the specific instance of `ValueWidth` such
+    /// that the enum discriminant values matches ICU4C's enum integer value.
+    const ENUM_VALUE: ValueWidthEnum;
+    /// This value is used to indicate an error in the Rust code in accessing
+    /// a position in the trie's `data` array. In normal cases, the position in
+    /// the `data` array will return either the correct value, or in case of a
+    /// logical error in the trie's computation, the trie's own error value
+    /// which is stored that in the `data` array.
+    const ERROR_VALUE_RUST: Self;
```

Comment:
Nit (optional):

```suggestion
    const ERROR_VALUE: Self;
```

Reply:
I changed the name to something more specific to allow ERROR_VALUE to correspond to the ICU meaning of the value stored in the data array.